src/term/mod.rs

```diff
@@ -80,6 +96,52 @@ impl selection::SemanticSearch for Term {
 
         point
     }
+
+    fn url_search(&self, mut point: Point<usize>) -> Option<String> {
+        point.line = self.grid.num_lines().0 - point.line - 1;
+
+        // Limit the starting point to the last line in the history
+        point.line = min(point.line, self.grid.len() - 1);
+
+        // Create forwards and backwards iterators
+        let iterf = self.grid.iter_from(point);
+        point.col += 1;
+        let mut iterb = self.grid.iter_from(point);
+
+        // Put all characters until separators into a string
+        let mut buf = String::new();
+        while let Some(cell) = iterb.prev() {
+            if URL_SEPARATOR_CHARS.contains(&cell.c) {
+                break;
+            }
+            buf.insert(0, cell.c);
+        }
+        for cell in iterf {
+            if URL_SEPARATOR_CHARS.contains(&cell.c) {
+                break;
+            }
+            buf.push(cell.c);
+        }
+
+        // Heuristic to remove all leading '('
+        while buf.starts_with('(') {
+            buf.remove(0);
+        }
```

Comment:
This seems quite inefficient to me. `remove(0)` is `O(n)`. Wouldn't it make sense to use `trim_left_matches()`?

Reply:
I wasn't much concerned with efficiency here, since it's only executed on click. But if this is re-used from a different code, that seems like a worthy cause.

I've looked at the `trim_left_matches` function, since I wasn't aware that exists, and gave it a quick spin just to confirm that it's faster. However I noticed that it is not.

If we're just comparing str trimming to str trimming, the `trim_left_matches` function is faster. However, we're trimming a `String`, not a `&str`. And the function returns a `&'a str`. So to make it work we'd have to change it to `trim_left_matches('(').into()`, which is super inefficient because of the `into`.

Here's the code I've used if you want to benchmark it yourself:
```
#![feature(test)]

extern crate test;

#[bench]
fn tlm(b: &mut test::Bencher) {
    b.iter(|| {
        let mut test = String::from("((Hello, World!");
        test = test.trim_left_matches('(').to_owned();
        test::black_box(test);
    });
}

#[bench]
fn other(b: &mut test::Bencher) {
    b.iter(|| {
        let mut test = String::from("((Hello, World!");
        while test.starts_with('(') {
            test.remove(0);
        }
        test::black_box(test);
    });
}
```

And these are the results I've got:
```
running 2 tests
test other ... bench:          26 ns/iter (+/- 0)
test tlm   ... bench:          42 ns/iter (+/- 0)
```

Note that I'm not saying that `remove(0)` is the most efficient way to do it here. I think it should be fine since we're usually trimming 0 characters, a realistic worse case might be 2-3. So checking first if any trimming is required at all probably resolves most performance issues.

If you have any other idea how to make this fast **and** easy to read, let me know. For now I think this should be an appropriate representation of what we're trying to achieve.

Reply:
Thanks for the exhaustive answer! I didn't consider the fact that it was an owned string and that the trimming function was defined for `str`. I also thought that it was more expensive for `remove(0)` to move the whole string.